src/terms/xc.jl

```diff
@@ -163,42 +163,51 @@ end
     isnothing(term.ρcore) && return nothing
 
     Vxc_real = xc_potential_real(term, basis, ψ, occupation; ρ, τ).potential
-    # TODO: move arrays to the CPU to enable forces calculations on the GPU.
-    #       Might require optimizations in the future.
     if basis.model.spin_polarization in (:none, :spinless)
-        Vxc_fourier = to_cpu(fft(basis, Vxc_real[:,:,:,1]))
+        Vxc_fourier = fft(basis, Vxc_real[:,:,:,1])
     else
-        Vxc_fourier = to_cpu(fft(basis, mean(Vxc_real, dims=4)))
+        Vxc_fourier = fft(basis, mean(Vxc_real, dims=4))
     end
 
     model = basis.model
-    form_factors = atomic_density_form_factors(basis, CoreDensity())
+    form_factors, iG2ifnorm = atomic_density_form_factors(basis, CoreDensity())
     nlcc_groups = [(igroup, group) for (igroup, group) in enumerate(basis.model.atom_groups)
                    if has_core_density(model.atoms[first(group)])]
     @assert !isnothing(nlcc_groups)
 
+    # Pre-allocation of large arrays for GPU efficiency
     TT = promote_type(T, eltype(Vxc_real))
+    Gs = G_vectors(basis)
+    work = to_device(basis.architecture, zeros(Complex{TT}, length(Gs)))
+    indices = to_device(basis.architecture, collect(1:length(Gs)))
+
     forces = [zero(Vec3{TT}) for _ = 1:length(model.positions)]
     for (igroup, group) in nlcc_groups
         for iatom in group
             r = model.positions[iatom]
-            forces[iatom] = _force_xc(basis, Vxc_fourier, form_factors, igroup, r)
+            forces[iatom] = _force_xc(basis, Vxc_fourier, form_factors, iG2ifnorm,
+                                      igroup, r, Gs, work, indices)
         end
     end
     forces
 end
 
 # Function barrier to work around various type instabilities.
```

Comment:
Perhaps what you did solves the type instabilities. Can you check that ?

In case it does we can inline the function, making this less clunky.

Reply:
Unfortunately, the type instability is still present. I changed the `_force_xc()` function to include the loop over atoms and make its call less clunky